File: libs/adsp-service-sdk/src/event/eventService.ts
import axios from 'axios';
import { Logger } from 'winston';
import { TokenProvider } from '../access';
import { ServiceDirectory } from '../directory';
import { AdspId, adspId } from '../utils';
import type { DomainEvent, DomainEventDefinition } from './event';

/**
 * Interface to the event service for sending domain events.
 *
 * @export
 * @interface EventService
 */
export interface EventService {
  /**
   * Sends a domain event of the service.
   *
   * Domain event sent uses the service name as its namespace.
   *
   * @param {DomainEvent} event
   * @memberof EventService
   */
  send(event: DomainEvent): void;
}

export class EventServiceImpl implements EventService {
  private readonly LOG_CONTEXT = { context: 'EventService' };

  private readonly namespace: string;
  private readonly definitions: string[];

  constructor(
    private readonly isCore: boolean,
    private readonly logger: Logger,
    private readonly directory: ServiceDirectory,
    private readonly tokenProvider: TokenProvider,
    serviceId: AdspId,
    events: DomainEventDefinition[]
  ) {
    this.namespace = serviceId.service;
    this.definitions = events?.map((e) => e.name) || [];
  }

  async send({ tenantId, ...event }: DomainEvent): Promise<void> {
    if (!this.definitions.includes(event.name)) {
      throw new Error(`Event ${this.namespace}:${event.name} is not recognized; only registered events can be sent.`);
    }

    const serviceUrl = await this.directory.getServiceUrl(adspId`urn:ads:platform:event-service:v1`);
    const sendUrl = new URL('v1/events', serviceUrl);

    try {
      const token = await this.tokenProvider.getAccessToken();

      this.logger.debug(`Sending event ${this.namespace}:${event.name} to: ${sendUrl}...`, {
        ...this.LOG_CONTEXT,
        tenant: tenantId?.toString(),
      });

      await axios.post(
        sendUrl.href,
        this.isCore
          ? { ...event, namespace: this.namespace, tenantId: `${tenantId}` }
          : { ...event, namespace: this.namespace },
        { headers: { Authorization: `Bearer ${token}` } }
      );

      this.logger.info(`Sent domain event ${this.namespace}:${event.name}.`, {
        ...this.LOG_CONTEXT,
        tenant: tenantId?.toString(),
      });
    } catch (err) {
      this.logger.error(`Error encountered on sending of event ${this.namespace}:${event.name}. ${err}`, {
        ...this.LOG_CONTEXT,
        tenant: tenantId?.toString(),
      });
    }
  }
}

File: libs/adsp-service-net-sdk/Events/EventService.cs
using System.Diagnostics.CodeAnalysis;
using System.Net;
using Adsp.Sdk.Registration;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using RestSharp;

namespace Adsp.Sdk.Events;
[SuppressMessage("Usage", "CA1812: Avoid uninstantiated internal classes", Justification = "Instantiated by dependency injection")]
internal sealed class EventService : IEventService, IDisposable
{
  private static readonly AdspId EVENT_SERVICE_API_ID = AdspId.Parse("urn:ads:platform:event-service:v1");

  private readonly ILogger<EventService> _logger;
  private readonly IServiceRegistrar _registrar;
  private readonly IServiceDirectory _serviceDirectory;
  private readonly ITokenProvider _tokenProvider;
  private readonly string _namespace;
  private readonly IRestClient _client;

  public EventService(
    ILogger<EventService> logger,
    IServiceRegistrar registrar,
    IServiceDirectory serviceDirectory,
    ITokenProvider tokenProvider,
    IOptions<AdspOptions> options,
    RestClient? client = null
  )
  {
    if (options.Value.ServiceId?.Service == null)
    {
      throw new ArgumentException("Provided options must ADSP service URN value for ServiceId.");
    }

    _logger = logger;
    _registrar = registrar;
    _serviceDirectory = serviceDirectory;
    _tokenProvider = tokenProvider;
    _namespace = options.Value.ServiceId.Service;
    _client = client ?? new RestClient();
  }

  public async Task Send<TPayload>(DomainEvent<TPayload> @event, AdspId? tenantId = null) where TPayload : class
  {
    var definition = _registrar.GetEventDefinition(@event.Name);
    if (definition == null)
    {
      throw new InvalidOperationException(
        $"Specified event '{@event.Name}' is not recognized. Event definition must be registered."
      );
    }

    var eventServiceUrl = await _serviceDirectory.GetServiceUrl(EVENT_SERVICE_API_ID);
    var requestUrl = new Uri(eventServiceUrl, "v1/events");

    var token = await _tokenProvider.GetAccessToken();
    var request = new RestRequest(requestUrl, Method.Post);
    request.AddHeader("Authorization", $"Bearer {token}");
    request.AddJsonBody(new FullDomainEvent<TPayload>(_namespace, @event, tenantId));

    var response = await _client.ExecuteAsync(request);
    if (response.StatusCode != HttpStatusCode.OK)
    {
      _logger.LogError(
        "Error encountered sending domain event with status '{Status}': {Content}",
        response.StatusCode,
        response.Content
      );
    }
  }

  public void Dispose()
  {
    _client.Dispose();
  }
}
